fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/item/InventorySlotWrapper.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.transfer.item;
+
+import net.minecraft.inventory.Inventory;
+import net.minecraft.item.ItemStack;
+
+import net.fabricmc.fabric.api.transfer.v1.item.InventoryStorage;
+import net.fabricmc.fabric.api.transfer.v1.item.ItemVariant;
+import net.fabricmc.fabric.api.transfer.v1.storage.StoragePreconditions;
+import net.fabricmc.fabric.api.transfer.v1.storage.base.SingleSlotStorage;
+import net.fabricmc.fabric.api.transfer.v1.transaction.TransactionContext;
+import net.fabricmc.fabric.api.transfer.v1.transaction.base.SnapshotParticipant;
+
+/**
+ * A wrapper around a single slot of an inventory
+ * We must ensure that only one instance of this class exists for every inventory slot,
+ * or the transaction logic will not work correctly.
+ * This is handled by the Map in InventoryWrappersImpl.
+ */
+class InventorySlotWrapper extends SnapshotParticipant<ItemStack> implements SingleSlotStorage<ItemVariant> {
+	/**
+	 * Strong reference to the InventoryStorage array to ensure that the weak values don't get GC'ed when InventoryStorages are still being accessed.
+	 */
+	@SuppressWarnings({"FieldCanBeLocal", "unused"})
+	private final InventoryStorage[] strongRef;
+	final Inventory inventory;
+	final int slot;
+
+	InventorySlotWrapper(Inventory inventory, int slot, InventoryStorage[] strongRef) {
+		this.inventory = inventory;
+		this.slot = slot;
+		this.strongRef = strongRef;
+	}
+
+	@Override
+	public long insert(ItemVariant key, long maxAmount, TransactionContext transaction) {
+		StoragePreconditions.notBlank(key);
+		int count = (int) Math.min(maxAmount, inventory.getMaxCountPerStack());
+		ItemStack stack = inventory.getStack(slot);
+
+		if (stack.isEmpty()) {
+			ItemStack keyStack = key.toStack(count);
+
+			if (inventory.isValid(slot, keyStack)) {
+				int inserted = Math.min(keyStack.getMaxCount(), count);
+				this.updateSnapshots(transaction);
+				keyStack.setCount(inserted);
+				inventory.setStack(slot, keyStack);
+				return inserted;
+			}
+		} else if (key.matches(stack)) {
+			int inserted = Math.min(stack.getMaxCount() - stack.getCount(), count);
```

Comment:
Merging stacks here doesn't account for `inventory.getMaxCountPerStack()`

Reply:
```java
int count = (int) Math.min(maxAmount, inventory.getMaxCountPerStack());
```
at the beginning of the function accounts for it!

Reply:
Well no that was wrong in case some items were already in the slot, however my new commit fixes this! Thanks!